File: core/src/main/java/com/caibowen/prma/core/filter/basic/Filter.java
package com.caibowen.prma.core.filter.basic;

import com.caibowen.gplume.annotation.NoExcept;
import com.caibowen.prma.core.LifeCycle;

/**
 * @author BowenCai
 * @since 29-10-2014.
 */
public interface Filter<E> extends LifeCycle {

    /**
     *
     * @param e
     * @return -1 for reject, 1 for accept, 0 for neutral.
     */
    @NoExcept
    int accept(E e);

    Filter getNext();

    void setNext(Filter ef);

}

File: core/src/main/scala/com/caibowen/prma/core/filter/Filter.scala
package com.caibowen.prma.core.filter

import com.caibowen.prma.core.LifeCycle

/**
* @author BowenCai
* @since  01/12/2014.
*/
trait Filter[T] extends LifeCycle {

  var next: Filter[T] = _

  @inline
  protected def doAccept(v: T): Int = -1

  def accept(v: T): Int = {
    if (1 == doAccept(v)) {
      if (next != null)
        next.accept(v)
      else 1

    } else -1

  }
}
